scripts/kconfig/kconfig.py

```diff
@@ -119,27 +115,67 @@ def check_no_promptless_assign(kconf):
 """ + SYM_INFO_HINT).format(sym))
 
 
-def check_assigned_sym_value(sym):
-    # Verifies that the value assigned to 'sym' "took" (matches the value the
-    # symbol actually got), printing a warning otherwise
+def check_assigned_sym_values(kconf):
+    # Verifies that the values assigned to symbols "took" (matches the value
+    # the symbols actually got), printing warnings otherwise. Choice symbols
+    # are checked separately, in check_assigned_choice_values().
+
+    for sym in kconf.unique_defined_syms:
+        if sym.choice:
+            continue
 
-    # Tristate values are represented as 0, 1, 2. Having them as
-    # "n", "m", "y" is more convenient here, so convert.
-    if sym.type in (BOOL, TRISTATE):
-        user_value = TRI_TO_STR[sym.user_value]
-    else:
         user_value = sym.user_value
+        if user_value is None:
+            continue
 
-    if user_value != sym.str_value:
-        warn(("""\
-{0.name_and_loc} was assigned the value '{1}' but got the value
-'{0.str_value}'. Check its dependencies. \
-""" + SYM_INFO_HINT).format(sym, user_value))
+        # Tristate values are represented as 0, 1, 2. Having them as "n", "m",
+        # "y" is more convenient here, so convert.
+        if sym.type in (BOOL, TRISTATE):
+            user_value = TRI_TO_STR[user_value]
 
+        if user_value != sym.str_value:
+            msg = f"{sym.name_and_loc} was assigned the value '{user_value}' " \
+                  f"but got the value '{sym.str_value}'. "
 
-def check_assigned_choice_value(choice):
-    # Verifies that the choice symbol that was selected (by setting it to y)
-    # ended up as the selection, printing a warning otherwise.
+            # List any unsatisfied 'depends on' dependencies in the warning
+            mdeps = missing_deps(sym)
+            if mdeps:
+                msg += "Check these unsatisfied dependencies: " + \
+                    ", ".join(f"{expr_str(expr)} (={TRI_TO_STR[expr_value(expr)]})"
+                              for expr in mdeps)
+                msg += ". "
+
+            warn(msg + SYM_INFO_HINT.format(sym))
+
+
+def missing_deps(sym):
+    # check_assigned_sym_values() helper for finding unsatisfied dependencies.
+    #
+    # Given direct dependencies
+    #
+    #     depends on <expr> && <expr> && ... && <expr>
+    #
+    # on 'sym' (which can also come from e.g. a surrounding 'if'), returns a
+    # list of all <expr>s with a value less than the value 'sym' was assigned
+    # ("less" instead of "not equal" just to be general and handle tristates,
+    # even though Zephyr doesn't use them).
```

Comment:
"a value less than", does this means we won't be warned of an unsatisfied symbol when we have

`depends on ! A
`
?

Reply:
Ah, the expr itself is less than. This should be OK then.

Reply:
Does this behave sanely when there are multiple config declarations with different dependencies, thereby creating a "super-config" with each configs dependencies OR'd together?

Reply:
It will work at least, though it might require some more manual investigation depending on how the dependencies look.

Might get `Check these unsatisfied dependencies: FOO || BAR || BAZ (=n), ...` for example. The code always just splits on AND here.

Reply:
Forgot to mention: `Symbol.direct_dep` has the OR of all the dependencies when a symbol is defined in multiple locations.

Reply:
> Ah, the expr itself is less than. This should be OK then.

Yeah, the way it avoids ever warning for `CONFIG_FOO=n` is because the dependencies can never be less than `n`. Avoids special-casing it.

Reply:
Great.

Reply:
Done. Now puts `()` around expressions that aren't plain symbols:

```
Check these unsatisfied dependencies: BAR (=n), BAZ (=n), (BAZ || BAZ) (=n), (STR = "hmm") (=n)
```

(Added a `BAZ || BAZ` dependency just to test.)

Reply:
As a side note, it'll always be `(=n)` in Zephyr, but it could be `(=m)` in projects that use `tristate` symbols (used for modules in Linux). Doesn't hurt to show it at least.